zerver/views/messages.py

```diff
@@ -1582,11 +1583,63 @@ def update_message_backend(request: HttpRequest, user_profile: UserMessage,
 
         mention_user_ids = message.mentions_user_ids
 
-    number_changed = do_update_message(user_profile, message, topic_name,
-                                       propagate_mode, content, rendered_content,
+    new_stream = None
+    old_stream = None
+    number_changed = 0
+
+    if new_stream_id is not None:
+        if not user_profile.is_realm_admin:
+            raise JsonableError(_("You don't have permission to move this message"))
+        if content is not None:
+            raise JsonableError(_("Cannot change message content while changing stream"))
+        if topic_name is not None:
+            # TODO: Add support for changing the topic together with
+            # the stream; it seems useful and should easily fit within
+            # an extended topic-edit UI.
+            raise JsonableError(_("Cannot change topic while changing stream"))
+
+        old_topic_name = message.topic_name()
+
+        old_stream = get_stream_by_id(message.recipient.type_id)
+        new_stream = get_stream_by_id(new_stream_id)
+
+        if not (old_stream.is_public() and new_stream.is_public()):
+            # We'll likely decide to relax this condition in the
+            # future; it just requires more care with details like the
+            # breadcrumb messages.
+            raise JsonableError(_("Streams must be public"))
+
+    number_changed = do_update_message(user_profile, message, new_stream,
+                                       topic_name, propagate_mode,
+                                       content, rendered_content,
                                        prior_mention_user_ids,
                                        mention_user_ids, mention_data)
 
+    if (number_changed > 0 and new_stream_id is not None) and (
+        new_stream is not None and old_stream is not None
+    ):
+        # Since moving content between streams is highly disruptive,
+        # it's worth adding a couple tombstone messages showing what
+        # happened.
+        sender = get_system_bot(settings.NOTIFICATION_BOT)
+
+        # Send a notification to the new stream that the topic was moved.
+        user_mention = "**_%s|%s**" % (user_profile.full_name, user_profile.id)
+        old_topic_link = "#**%s>%s**" % (old_stream.name, old_topic_name)
+
+        if topic_name is None:
+            topic_name = old_topic_name
+
+        internal_send_stream_message(
+            new_stream.realm, sender, new_stream, topic_name,
+            "This topic was moved here from %s by %s" % (old_topic_link, user_mention))
+
+        # Send a notification to the old stream that the topic was moved.
+        new_topic_link = "#**%s>%s**" % (new_stream.name, topic_name)
+        internal_send_stream_message(
+            old_stream.realm, sender, old_stream, old_topic_name,
+            "This topic was moved by %s to %s" % (user_mention, new_topic_link))
```

Comment:
We should make sure that one of your tests verifies the `message.content` for these two messages.

Additionally, we should tag these strings for translation using `_()`.